src/Cardano/Wallet/Kernel/Wallet.hs

```diff
@@ -0,0 +1,119 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE TypeOperators #-}
+
+-- |
+-- Copyright: © 2018-2019 IOHK
+-- License: MIT
+--
+-- | Module provides internal implementation of
+-- | wallet api
+
+module Cardano.Wallet.Kernel.Wallet where
+
+import Prelude
+
+import Cardano.Wallet
+    ( Wallet )
+import Cardano.Wallet.AddressDerivation
+    ( Passphrase )
+import Cardano.Wallet.AddressDiscovery
+    ( AddressPoolGap )
+import Cardano.Wallet.Mnemonic
+    ( Mnemonic )
+import Cardano.Wallet.Primitive
+    ( Coin )
+import Crypto.Encoding.BIP39
+    ( ValidMnemonicSentence )
+import Data.Text
+    ( Text )
+import Data.Time.Units
+    ( Microsecond )
+import Fmt
+    ( Buildable (..), fmt )
+
+-- | Errors
+data CreateWalletError = CreateWalletError
+
+instance Buildable CreateWalletError where
+    build CreateWalletError =
+        mempty
+        <> "CreateWalletError"
+
+instance Show CreateWalletError where
+    show = fmt . build
+
+data GetWalletError = GetWalletError
+
+instance Buildable GetWalletError where
+    build GetWalletError =
+        mempty
+        <> "GetWalletError"
+
+instance Show GetWalletError where
+    show = fmt . build
+
+
+-- | Types
+newtype CreateWallet = CreateWallet NewWallet
+
+data NewWallet =
+    forall mw . ValidMnemonicSentence mw => NewWallet
+    {
+      _mnemonicSentence
+        :: !(Mnemonic mw)
+    , _mnemonicSentencePassphrase
+        :: !(Maybe (Passphrase "generation"))
+    , _name
+        :: !WalletName
+    , _passphrase
+        :: !(Passphrase "encryption")
+    , _addressPoolGap
+        :: !AddressPoolGap
+    }
+
+newtype WalletName = WalletName { _getWalletName :: Text }
+    deriving (Eq, Show)
+
+newtype WalletId = WalletId Text
+    deriving (Eq, Show)
+
+newtype WalletTimestamp = WalletTimestamp Microsecond
+    deriving (Eq, Ord, Show)
+
+data WalletOperation = Ready | Restoring deriving (Eq, Show)
+
+data Delegation = Delegated | NotDelegated deriving (Eq, Show)
+
+newtype PassphraseInfo = PassphraseInfo { lastUpdated :: WalletTimestamp }
+    deriving (Eq, Show)
+
+data WalletState = WalletState {
+      id
+        :: !WalletId
+    , name
+        :: !WalletName
+    , balance
+        :: !Coin
```

Comment:
This, we compute from the wallet UTxOs directly :+1:, no need to store or represent that at this level.

Reply:
removed the `balance` field